File: leetcode-algorithms/211. Add and Search Word - Data structure design/Solution.cpp
//
// Design a data structure that supports the following two operations:
//
// void addWord(word)
// bool search(word)
// search(word) can search a literal word or a regular expression string containing only letters a-z or .. A . means it can represent any one letter.
//
// Example:
//
// addWord("bad")
// addWord("dad")
// addWord("mad")
// search("pad") -> false
// search("bad") -> true
// search(".ad") -> true
// search("b..") -> true
// Note:
// You may assume that all words are consist of lowercase letters a-z.

class TrieNode {
public:
    bool bleafe;
    vector<TrieNode*> children;
    TrieNode() : bleafe(false), children(26, nullptr) {}
};

class Trie {
public:
    /** Initialize your data structure here. */
    Trie() {
        root = new TrieNode();
    }

    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode* ro = root;
        for (auto c : word) {
            if (ro->children[c - 'a'] == nullptr) {
                ro->children[c - 'a'] = new TrieNode();
            }
            ro = ro->children[c - 'a'];
        }
        ro->bleafe = true;
    }

    /** Returns if the word is in the trie. */
    bool search(string word) {
        return searchFromNode(root, word, 0);
    }

    bool searchFromNode(TrieNode* node, string& word, int pos) {
        if (node == nullptr) return false;
        if (word.size() == pos) return node->bleafe;
        if (word[pos] == '.') {
            for (auto child : node->children) {
                if (searchFromNode(child, word, pos + 1)) return true;
            }
            return false;
        } else {
            if (node->children[word[pos] - 'a'] == nullptr) return false;
            else return searchFromNode(node->children[word[pos] - 'a'], word, pos + 1);
        }
    }

    TrieNode* root;
};

class WordDictionary {
public:
    /** Initialize your data structure here. */
    WordDictionary() {

    }

    /** Adds a word into the data structure. */
    void addWord(string word) {
        trie.insert(word);
    }

    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    bool search(string word) {
        return trie.search(word);
    }

    Trie trie;
};


/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * bool param_2 = obj.search(word);
 */

File: leetcode-algorithms/211. Add and Search Word - Data structure design/Solution.java
class TrieNode {

    public TrieNode() {
        isend = false;
        children = new TrieNode[26];
        for (int i = 0; i < 26; i++) children[i] = null;
    }
    boolean isend;
    TrieNode[] children;
}

class Trie {

    /** Initialize your data structure here. */
    public Trie() {

    }

    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode node = root;

        for (int i = 0; i < word.length(); i++) {
            int idx = (int)(word.charAt(i) - 'a');
            if (node.children[idx] == null) {
                node.children[idx] = new TrieNode();
            }

            node = node.children[idx];
        }

        node.isend = true;
    }

    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode node = root;
        for (int i = 0; i < word.length(); i++) {
            int idx = (int)(word.charAt(i) - 'a');

            if (node.children[idx] == null) return false;
            else {
                node = node.children[idx];
            }
        }
        return node.isend;
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode node = root;
        for (int i = 0; i < prefix.length(); i++) {
            int idx = (int)(prefix.charAt(i) - 'a');

            if (node.children[idx] == null) return false;
            else {
                node = node.children[idx];
            }
        }
        return true;
    }

    TrieNode root = new TrieNode();
}

class WordDictionary {

    /** Initialize your data structure here. */
    public WordDictionary() {

    }

    /** Adds a word into the data structure. */
    public void addWord(String word) {
        trie.insert(word);
    }

    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    public boolean search(String word) {
        return implsearch(word, trie.root);
    }

    private boolean implsearch(String word, TrieNode node) {
        if (word.length() == 0) return (node != null) && (node.isend == true);

        char c = word.charAt(0);
        if (c != '.') {
            int idx = (int) (c - 'a');
            if (node.children[idx] == null) return false;
            return implsearch(word.substring(1), node.children[idx]);
        } else {
            for (TrieNode child : node.children) {
                if (child != null && implsearch(word.substring(1), child)) return true;
            }
            return false;
        }
    }

    Trie trie = new Trie();
}
